sapicore/engine/network/__init__.py

```diff
@@ -0,0 +1,400 @@
+
+    def _out_edges(self, node: str) -> list[Synapse]:
+        """Returns list of synapse objects fanning out of the ensemble `node`."""
+        return [ref.get("reference") for _, _, ref in list(self.graph.out_edges(node, data=True))]
+
+    def _find_roots(self):
+        """Identifies nodes that are exposed to external input. Those are used as starting points for the multi-BFS
+        forward sweep, whose order also gets updated every time this method is called.
+
+        Warning
+        -------
+        Roots are updated on network initialization and on every application of :meth:`add_ensemble` and
+        :meth:`add_synapse`. If adding a synapse results in a rootless recurrent network, all nodes will have
+        in-degree > 0. In such cases, the method will mark the first ensemble added `self.get_ensembles()[0]`
+        as the root and direct external input to it. This behavior can be modified by setting `self.roots`
+        directly after building the network.
+
+        """
+        updated_roots = [node for node, degree in self.graph.in_degree() if degree == 0]
+        if updated_roots:
+            self.roots = updated_roots
+
+        elif self.get_ensembles():
+            # if the network became rootless after adding a synapse, default to the first ensemble as the root.
+            self.roots = [self.get_ensembles()[0].identifier]
+
+        # breadth-first search (BFS) iterator used to traverse the network graph layer by layer.
+        # returns a list of layers, first one consisting of the root nodes. Flatten and visit sequentially.
+        self.traversal_order = flatten(list(nx.bfs_layers(self.graph, self.roots)))
+
+    def _object_from_configuration(
+        self, path: str = None, cfg: dict = None, comp_type: str = None
+    ) -> (Neuron | Synapse, dict):
+        """Constructs an objects from the YAML in `path` or from `cfg` dictionary if provided."""
+        # load configuration dictionary from ensemble descriptor YAML.
+        if path and not cfg:
+            cfg = load_yaml(path)
+
+        # initialize the ensemble/synapse based on its class reference, given in its configuration file.
+        temp_import = __import__(cfg["package"], globals(), locals(), ["*"])
+        class_ref = getattr(temp_import, cfg["class"])
+
+        if "ensemble" in comp_type:
+            comp_ref = class_ref(**cfg["model"], configuration=cfg, device=self.device)
+
+        else:
+            comp_ref = class_ref(
+                **cfg["model"],
+                configuration=cfg,
+                device=self.device,
+                src_ensemble=self.graph.nodes[cfg["model"]["source"]]["reference"],
+                dst_ensemble=self.graph.nodes[cfg["model"]["target"]]["reference"],
+            )
+
+        # pass autograd setting from network configuration file to the synapse, if provided.
+        if cfg.get("autograd"):
+            cfg["model"]["autograd"] = self.cfg["autograd"]
```

Comment:
Just wondering why `autograd` is special :)

Reply:
There doesn't seem to be a reason for it to be passed separately. That snippet was added back when Jeremy was concerned about supporting backward passes. I bet you'll find more ancient artifacts in there :)